### PR TITLE
Improve handling of user header

### LINKED ISSUE #812: user_header will not enter the $cpp_options() if the model is not recompiled
**Describe the bug**
The contents of `CmdStanModel$cpp_options()` depend on whether a user header was passed via tha `user_header` argument or explicitly via tha `cpp_options` argument

**To Reproduce**
```r
# Make sure target file does not exist
file.remove("gp.exe")
# Compile 
m <- cmdstan_model("gp.stan", stanc_options = list("allow-undefined"), user_header = here::here("gp.hpp"))
m$cpp_options() #Includes USER_HEADER

# No change, just load the model
m <- cmdstan_model("gp.stan", stanc_options = list("allow-undefined"), user_header = here::here("gp.hpp"))
m$cpp_options() #Does not include USER_HEADER

m <- cmdstan_model("gp.stan", stanc_options = list("allow-undefined"), user_header = here::here("gp.hpp"), force_recompile = TRUE)
m$cpp_options() #Includes USER_HEADER


# No change, just load the model
m <- cmdstan_model("gp.stan", stanc_options = list("allow-undefined"),  cpp_options = list(USER_HEADER = here::here("gp.hpp")))
m$cpp_options() #Includes USER_HEADER
```

**Expected behavior**
I'd expect `USER_HEADER` to be a part of `cpp_options()` in all cases

**Operating system**
Windows 10

**CmdStanR version number**
0.6.0

### PR DESCRIPTION
**Pull request by martinmodrak**
#### Submission Checklist

- [x] Run unit tests
- [x] Declare copyright holder and agree to license (see below)

#### Summary

Resolves #812 and #813 (as the fix to #813 requires changes needed to fix #812, I've made a single PR for both)

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting
(this will be you or your assignee, such as a university or company):
Institute of Microbiology of the Czech Academy of Sciences

By submitting this pull request, the copyright holder is agreeing to
license the submitted work under the following licenses:

- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by martinmodrak** [comment]
@rok-cesnovar @jgabry  So this appears ready to merge. Beyond the changes to header handling I made some tweaks to tests. In particular, there are now `expect_compile` and `expect_no_recompilation` helpers that standardize the behaviour to have file existence and modifications time checked at all places and not rely only on messages (which I am not sure are actually checked when testing within the CI environment as they are bound to interactive mode). I've also added a `dry_run` argument to `$compile` that let's us test argument handling and messages without actually trigerring the C++ compilation. Both could probably be used more extensively, but I didn't want to pollute the PR too much. If you like the concept, I will open a separate issue + PR on applying this more broadly.

File: R/model.R [hunk 1]
```diff
@@ -397,6 +397,8 @@ CmdStanModel <- R6::R6Class(
 #'   `functions` field in the compiled model object. This can also be done after
 #'   compilation using the
 #'   [`$expose_functions()`][model-method-expose_functions] method.
+#' @param dry_run (logical) If TRUE, the code will do all checks before compilation,
```
Comment (jgabry) [inline R/model.R:400]:
```suggestion
#' @param dry_run (logical) If `TRUE`, the code will do all checks before compilation,
```

File: R/model.R [hunk 2]
```diff
@@ -500,15 +504,63 @@ compile <- function(quiet = TRUE,
...
+  }
+
+  if (isTRUE(cpp_options$stan_opencl)) {
+    stanc_options[["use-opencl"]] <- TRUE
+  }
+
+  # Note that unlike cpp_options["USER_HEADER"], the user_header variable is deliberately
+  # not transformed with wsl_safe_path() as that breaks the check below on WSLv1
+  if (!is.null(user_header)) {
+    if(!is.null(cpp_options[["USER_HEADER"]]) || !is.null(cpp_options[["user_header"]])) {
```
Comment (jgabry) [inline R/model.R:519]:
```suggestion
    if (!is.null(cpp_options[["USER_HEADER"]]) || !is.null(cpp_options[["user_header"]])) {
```

File: R/model.R [hunk 3]
```diff
@@ -500,15 +504,63 @@ compile <- function(quiet = TRUE,
...
+      warning("User header specified both via user_header argument and via cpp_options arguments")
+    }
+
+    cpp_options[["USER_HEADER"]] <- wsl_safe_path(absolute_path(user_header))
+    stanc_options[["allow-undefined"]] <- TRUE
+    private$using_user_header_ <- TRUE
+  }
+  else if (!is.null(cpp_options[["USER_HEADER"]])) {
+    if(!is.null(cpp_options[["user_header"]])) {
+      warning('User header specified both via cpp_options[["USER_HEADER"]] and cpp_options[["user_header"]]')
```
Comment (jgabry) [inline R/model.R:529]:
```suggestion
      warning('User header specified both via cpp_options[["USER_HEADER"]] and cpp_options[["user_header"]].', call. = FALSE)
```

File: R/model.R [hunk 4]
```diff
@@ -500,15 +504,63 @@ compile <- function(quiet = TRUE,
...
+    user_header <- cpp_options[["user_header"]]
+    cpp_options[["user_header"]] <- wsl_safe_path(absolute_path(cpp_options[["user_header"]]))
+    private$using_user_header_ <- TRUE
+  }
+
+
+  if(!is.null(user_header)) {
+    user_header <- absolute_path(user_header) # As mentioned above, just absolute, not wsl_safe_path()
+    if(!file.exists(user_header)) {
+      stop(paste0("User header file '", user_header, "' does not exist"))
```
Comment (jgabry) [inline R/model.R:546]:
```suggestion
      stop(paste0("User header file '", user_header, "' does not exist."), call. = FALSE)
```

File: R/model.R [hunk 5]
```diff
@@ -529,7 +581,7 @@ compile <- function(quiet = TRUE,
 
   if (os_is_wsl() && (compile_model_methods || compile_standalone)) {
     warning("Additional model methods and standalone functions are not ",
-            "currently available with WSL CmdStan and will not be compiled",
+            "currently available with WSLv1 CmdStan and will not be compiled",
```
Comment (jgabry) [inline R/model.R:584]:
```suggestion
            "currently available with WSLv1 CmdStan and will not be compiled.",
```

**Review by jgabry — CHANGES_REQUESTED** [review]
Thanks @martinmodrak. I made a few suggestions but I'm just being nitpicky to match style, no substantive changes. @rok-cesnovar has written much more of the `compile` method code than I have so let's see if he has any other feedback, but this looks good to me.  

> I've also added a `dry_run` argument to `$compile` that let's us test argument handling and messages without actually trigerring the C++ compilation. 

Good idea.

> Both could probably be used more extensively, but I didn't want to pollute the PR too much. If you like the concept, I will open a separate issue + PR on applying this more broadly.

Great, yeah I like the concept.

**Review by rok-cesnovar — APPROVED** [review]
Thanks!

**Comment by martinmodrak** [comment]
Hi, just merged master back in. Could somebody merge this before it gets out of sync 🙏 ? Or is there something else to be done/approved?

**Comment by rok-cesnovar** [comment]
I just assume the author can merge once its approved :) Thanks Martin!

**Pull request merged**